# Solving a semi-infinite program (SIP) with equality constraints

[Matthew Wilhelm](https://psor.uconn.edu/person/matthew-wilhelm/)  
Department of Chemical and Biomolecular Engineering, University of Connecticut

### Using EAGO to solve a SIP w/ equality constraints

In general, the solution of semi-infinite programs with nonconvex semi-infinite equality in the form given below are extremely challenging. The general form of such problems is shown below:

$
\begin{align}\label{form:SIPImplicit}
    f^*=&\min_{\mathbf y}f(\mathbf{y}) \\
    &{\rm s.t.}\; \max_{\mathbf{p,z}}{g(\mathbf{z,y,p})}\le 0 \\
    &{\; \qquad}{\rm s.t.}\; \mathbf{h(z,y,p) =  0} \\
    & \mathbf{y} \in Y = \{y \in \mathbb{R} : \mathbf{y^{L}} \leq \mathbf{y} \leq \mathbf{y^{U}}\} \\
    & \mathbf{p} \in P = \{p \in \mathbb{R} : \mathbf{p^{L}} \leq \mathbf{p} \leq \mathbf{p^{U}}\} \\
    & \mathbf{z} \in D_{x} \subset \mathbb{R}^{n_{x}}
\end{align}
$

EAGO implements the impSIPres algorithm of [1] to determine a globally optimal solution to problems of the above form subject to the stiputation that $\mathbf{h(z,y,p) =  0}$ uniquely determines $y$ as an implicit function of $z$ and $p$ on $Y \times P \times D_x$. This accomplished using the `implicit_sip_solve` function which returns the optimal value, the solution, and a boolean feasibility value. To illustrate the functions use, a simple example is presented here:

$
\begin{align}
    f(y) &= (y-3.5)^4 - 5(y-3.5)^3 -(y-3.5)^2 + 30(y-3.5) \\
    h(z,y,p) &= z - (y - y^3/6 + y^5/120)/\sqrt{z} - p = 0 \\
    g(z,y,p) &= z + cos(y-p/90) - p \leq 0 \\
     &{\; \qquad}y \in Y = [0.5,80]\\
     &{\; \qquad}p \in P = [80,120]  \label{ex:equalSIP}
\end{align}
$

In [1]:
using EAGO

# Variable bounds
xl = [0.5]; xu = [8.0]; yl = [68.8]; yu = [149.9]
pl = [80]; pu = [120]; 

# Objective and constraint functions are defined
f(y) =  (y[1]-3.5)^4 - 5(y[1]-3.5)^3 - (y[1]-3.5)^2 + 30(y[1]-3.5)
gSIP(z,y,p) = z[1] + cos(y[1]-p[1]/90) - p[1]
function h!(out,z,y,p)
    out[1] = z[1] - (y[1] - y[1]^3/6 + y[1]^5/120)/sqrt(z[1]) - p[1]
end

# The impSIPres routine is run
low_val, upp_val, sol, feas = implicit_sip_solve(f, h!, gSIP, xl, xu, yl, yu, pl, pu)

println("The global minimum of the semi-infinite program is between: $low_val and $upp_val.")
println("The global minimum is attained at: y = $val.")

┌ Info: Precompiling EAGO [bb8be931-2a91-5aca-9f87-79e1cb69959a]
└ @ Base loading.jl:1192
┌ Warning: Package EAGO does not have CPLEX in its dependencies:
│ - If you have EAGO checked out for development and have
│   added CPLEX as a dependency but haven't updated your primary
│   environment's manifest file, try `Pkg.resolve()`.
│ - Otherwise you may need to report an issue with EAGO
└ Loading CPLEX into EAGO from project dependency, future warnings for EAGO are suppressed.
ERROR: LoadError: LoadError: could not open file C:\Users\wilhe\.julia\dev\EAGO\src\McCormick\ConvexityRules\convexity_rules.jl
Stacktrace:
 [1] include at .\boot.jl:317 [inlined]
 [2] include_relative(::Module, ::String) at .\loading.jl:1044
 [3] include at .\sysimg.jl:29 [inlined]
 [4] include(::String) at C:\Users\wilhe\.julia\dev\EAGO\src\McCormick\McCormick.jl:1
 [5] top-level scope at none:0
 [6] include at .\boot.jl:317 [inlined]
 [7] include_relative(::Module, ::String) at .\loading.jl:1044
 [8] include at 

ErrorException: Failed to precompile EAGO [bb8be931-2a91-5aca-9f87-79e1cb69959a] to C:\Users\wilhe\.julia\compiled\v1.0\EAGO\t0N0c.ji.

## References
1. M.D. Stuber and P.I. Barton, Semi-infinite optimization with implicit functions, *Industrial & Engineering Chemistry Research*, 54 (2015), pp. 307–317.